pkg/e2e/autoscaler/autoscaler.go

```diff
@@ -178,13 +178,87 @@ func machineAutoscalerResource(targetMachineSet *mapiv1beta1.MachineSet, minRepl
 
 var _ = g.Describe("[Feature:Machines] Autoscaler should", func() {
 	defer g.GinkgoRecover()
+	scaleUpFunc := func(client runtimeclient.Client, targetMachineSet mapiv1beta1.MachineSet, nodeTestLabel string, initialNumberOfReplicas int32, expectedReplicas int32) {
+		g.By(fmt.Sprintf("Creating MachineAutoscaler object for targetMachineSet %q", targetMachineSet.Name))
+		machineAutoscaler := machineAutoscalerResource(&targetMachineSet, 1, expectedReplicas)
+		err := client.Create(context.TODO(), machineAutoscaler)
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("Waiting for cluster to scale out number of replicas of targetMachineSet %q", targetMachineSet.Name))
+		err = wait.PollImmediate(5*time.Second, e2e.WaitLong, func() (bool, error) {
+			ms, err := e2e.GetMachineSet(context.TODO(), client, targetMachineSet.Name)
+			if err != nil {
+				glog.Errorf("error getting machineset object: %v, retrying...", err)
+				return false, nil
+			}
+			glog.Infof("MachineSet %q. Initial number of replicas: %d. Current number of replicas: %d", targetMachineSet.Name, initialNumberOfReplicas, pointer.Int32PtrDerefOr(ms.Spec.Replicas, e2e.DefaultMachineSetReplicas))
+			return pointer.Int32PtrDerefOr(ms.Spec.Replicas, e2e.DefaultMachineSetReplicas) == expectedReplicas, nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("Waiting for cluster to scale up nodes for targetMachineSet %q", targetMachineSet.Name))
+		err = wait.PollImmediate(5*time.Second, e2e.WaitLong, func() (bool, error) {
+			scaledMachines := mapiv1beta1.MachineList{}
+			if err := client.List(context.TODO(), runtimeclient.MatchingLabels(targetMachineSet.Spec.Selector.MatchLabels), &scaledMachines); err != nil {
+				glog.Errorf("Error querying api for machineset object: %v, retrying...", err)
+				return false, nil
+			}
+
+			// get all linked nodes and label them
+			nodeCounter := 0
+			for _, machine := range scaledMachines.Items {
+				if machine.Status.NodeRef == nil {
+					glog.Errorf("Machine %q does not have node reference set", machine.Name)
+					return false, nil
+				}
+				glog.Infof("Machine %q is linked to node %q", machine.Name, machine.Status.NodeRef.Name)
+				nodeCounter++
+			}
+
+			glog.Infof("Expecting %d nodes. Current number of nodes in the group: %d", expectedReplicas, nodeCounter)
+			return nodeCounter == int(expectedReplicas), nil
+		})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		g.By(fmt.Sprintf("Labeling all nodes belonging to the machineset %q (after scale up phase)", targetMachineSet.Name))
+		err = labelMachineSetNodes(client, &targetMachineSet, nodeTestLabel)
+		o.Expect(err).NotTo(o.HaveOccurred())
+	}
+
+	scaleDownFunc := func(client runtimeclient.Client, targetMachineSet string, nodeTestLabel string, initialNumberOfReplicas int32) {
```

Comment:
`scaleDownFunc` actually does not scale a machineset down. It just waits until a machineset is scaled down. Maybe `waitUntilScaledDownFunc`?